sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -281,42 +298,57 @@ public int hashCode() {
         }
         CompletableFuture<BitSet> sendRequestsResult = new CompletableFuture<>();
         AtomicInteger numRequests = new AtomicInteger(requestsByShard.size());
-        for (Iterator<Map.Entry<ShardLocation, TReq>> it = requestsByShard.entrySet().iterator(); it.hasNext(); ) {
+        Map<ShardLocation, TReq> batchRequests = new HashMap<>(requestsByShard);
+        requestsByShard.clear();
+
+        for (Iterator<Map.Entry<ShardLocation, TReq>> it = Iterables.cycle(batchRequests.entrySet()).iterator();
```

Comment:
by using this cycle pattern you will block the consuming thread until all items are removed. this is not what we want. afaik we defined that we will check for existing request limit exceedance on every node before we will executing any requests. if so, do nothing and stop consuming. what's the reason to come up with this different solution?